CombinatorialGames/Nimber/WWW.lean

```diff
@@ -0,0 +1,138 @@
+
+noncomputable section
+
+namespace Nimber
+
+protected def field_two : Subfield Nimber := Subfield.closure ∅
+
+private def tau : Nimber := of (omega0 ^ (omega0 ^ omega0))
+
+protected def degree (x : Nimber) : ℕ :=
+  (minpoly Nimber.field_two x).natDegree
+
+/- `kappa q` where `q = (nth Nat.Prime k) ^ (n + 1)` -/
+protected def kappa_q (k n : ℕ) : Nimber :=
+  of <| 2 ^ (omega0 ^ k * (nth Nat.Prime k) ^ n)
+
+protected def kappa (h : ℕ) : Nimber :=
+  sInf {x | x < tau ∧ h ∣ Nimber.degree x}
+
+protected def alpha (h : ℕ) : Nimber :=
+  sInf {x | x < Nimber.kappa h ∧ ∀ y < Nimber.kappa h, y.val ^ h ≠ x.val}
+
+protected theorem kappa_q_lt_iff_pair_lt (k n k' n' : ℕ) :
+  Nimber.kappa_q k n < Nimber.kappa_q k' n' ↔ k < k' ∨ k = k' ∧ n < n' := by
+  simp only [Nimber.kappa_q, OrderIso.lt_iff_lt]
+  rw [opow_lt_opow_iff_right (by simp)]
+  constructor
+  · contrapose!
+    intro ⟨h, h'⟩
+    rw [le_iff_lt_or_eq] at h
+    rcases h with h | rfl
+    · apply le_trans (b := omega0 ^ k)
+      · repeat rw [← opow_natCast]
+        rw [← natCast_opow]
+        apply le_of_lt
+        apply omega0_opow_mul_nat_lt
+        rwa [Nat.cast_lt]
+      · apply Ordinal.le_mul_left
+        rw [← opow_natCast]
+        apply opow_pos
+        rw [← Nat.cast_zero, Nat.cast_lt, Nat.pos_iff_ne_zero]
+        exact nth_prime_ne_zero k
+    · simp only [forall_const] at h'
+      rw [Ordinal.mul_le_mul_iff_left (by rw [← opow_natCast]; exact opow_pos _ omega0_pos)]
+      repeat rw [← opow_natCast, ← natCast_opow]
+      rwa [Nat.cast_le, Nat.pow_le_pow_iff_right (one_lt_nth_prime k')]
+  · intro h
+    rcases h with h | ⟨rfl, h⟩
+    · apply lt_of_lt_of_le (b := omega0 ^ k')
+      · repeat rw [← opow_natCast]
+        rw [← natCast_opow]
+        apply omega0_opow_mul_nat_lt
+        rwa [Nat.cast_lt]
+      · apply Ordinal.le_mul_left
+        rw [← opow_natCast]
+        apply opow_pos
+        rw [← Nat.cast_zero, Nat.cast_lt, Nat.pos_iff_ne_zero]
+        exact nth_prime_ne_zero k'
+    · rw [Ordinal.mul_lt_mul_iff_left (by rw [← opow_natCast]; exact opow_pos _ omega0_pos)]
+      repeat rw [← opow_natCast, ← natCast_opow]
+      rwa [Nat.cast_lt, Nat.pow_lt_pow_iff_right (one_lt_nth_prime k)]
+
+/-- Add two ordinals as nimbers. -/
+scoped notation:65 x:65 "+ₙᵢₘ" y:66 => val (∗x + ∗y)
+
+protected theorem base_two_expansion (x : Nimber) :
+  (CNF 2 x.val).foldr (fun p r ↦ 2 ^ p.1 * p.2 +ₙᵢₘ r) 0 = x.val := by
+  sorry
+
+protected theorem kappa_nth_prime_pow (k n : ℕ) {h : n ≠ 0} :
+  Nimber.kappa ((nth Nat.Prime k) ^ n) = Nimber.kappa_q k n := by
+  sorry
+
+protected theorem kappa_pow_eq_alpha (k : ℕ) :
+  (Nimber.kappa (nth Nat.Prime k)) ^ (nth Nat.Prime k) = Nimber.alpha (nth Nat.Prime k) := by
+  sorry
+
+protected theorem kappa_pow_decrease (k n : ℕ) {h : n ≠ 0} :
+  (Nimber.kappa ((nth Nat.Prime k) ^ (n + 1))) ^ (nth Nat.Prime k)
+  = (Nimber.kappa ((nth Nat.Prime k) ^ n)) := by
+  sorry
+
+private def tau' : Nimber :=
+  sInf {x | IsAlgClosed x}
```

Comment:
This is actually equal to `0` because of how we treat those edge cases. Oops.

To prove that `Nimber` is algebraically closed, I'll introduce a definition `algClosure : Nimber → Nimber`. You can then state `algClosure 2 = tau`.